Support RV32/RV64 mainline/metal stackings

timsifive opened the pull request:
Add an option that lets a user choose between metal and mainline FreeRTOS stackings (defaults to mainline).

Add support for RV64.

timsifive committed:
Support mainline FreeRTOS instead of metal FreeRTOS

I'll have to add an option or something before this can be merged.

The stack pointer value for suspended threads is computed, and I didn't
check that it's right. Can't be written.
Does not support accessing gp/tp in suspended threads.

Change-Id: Ibe7f5167b970d5990a296e968df2b4480135d673
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Add riscv_freertos_stacking command.

This lets the user describe how registers are stored on the stack.

Change-Id: I052188fa9c6cb4f8670fa4b01a8878272ed6fc4d

t-8ch commented:
FYI this works for me (FreeRTOS 10.3.1)!

At least after having adapted a few offsets, too:
```diff
@@ -176,12 +179,12 @@ static const struct FreeRTOS_params FreeRTOS_params_list[] = {
        "riscv",                        /* target_name */
        4,                                              /* thread_count_width; */
        4,                                              /* pointer_width; */
-       16,                                             /* list_next_offset; */
-       20,                                             /* list_width; */
+       4,                                              /* list_next_offset; */
+       24,                                             /* list_width; */
        8,                                              /* list_elem_next_offset; */
-       12,                                             /* list_elem_content_offset */
+       16,                                             /* list_elem_content_offset */
        0,                                              /* thread_stack_offset; */
-       52,                                             /* thread_name_offset; */
+       64,                                             /* thread_name_offset; */
```

t-8ch commented:
One issue that cropped up with this applied are with breakpoints.
Setting the breakpoint works correctly and it is correctly hit it will instantly rehit the same breakpoint over and over again.
Removing the breakpoint allows the code to continue.

timsifive commented:
@t-8ch, these offsets don't all make sense to me. How did you arrive at these new values? Did you use build options like configUSE_LIST_DATA_INTEGRITY_CHECK_BYTES that would affect the layout of these structures? (What is TickType_t in your project, and UBaseType_t?)

t-8ch commented:
@timsifive No that specifc value is not set, but other integrity options are.
( https://github.com/Ralim/IronOS/blob/master/source/Core/BSP/Pine64/Vendor/OS/FreeRTOS/Source/portable/GCC/portmacro.h )
I got these values manually by looking at the structure definitions in gdb.

I also wrote to the gdb mailing list to have a way of directly retrieving the offsets from the debug info via gdb:
https://sourceware.org/pipermail/gdb/2021-March/049329.html

timsifive committed:
Redo how we handle offsets in FreeRTOS.

Instead of hard-coding them for each "target," hard code the data types
in FreeRTOS list structures, and compute the offsets based on size of
pointers and ubase_t types.

Doesn't work right now.

Change-Id: I444cd1ef47121190e2222f19a67edf3c6155a96a

timsifive committed:
Correctly get thread list.

Works on RV32 and RV64.

Change-Id: I27768aef698475bef425d6a7e27ea609c9b9a1b6

timsifive committed:
Fix SP calculation and RV64 register stacking.

Smoketest now passes on spike with both RV32 and RV64.

Change-Id: I94b43e041abe5370a833bd3afb4a2a8591538d7a
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Style fixes.

Change-Id: I269b5aac8c233022c41ebc8ac8c5aeb437882719
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive committed:
Style fix.

Change-Id: I18fbff7dcaad9bd35f0942598c05c2a45bdb9f3b
Signed-off-by: Tim Newsome <tim@sifive.com>

GregSavin commented on src/rtos/FreeRTOS.c:
```diff
@@ -33,26 +33,13 @@
 
 struct FreeRTOS_params {
 	const char *target_name;
-	const unsigned char thread_count_width;
```
I'm curious of the rationale for moving these data members to struct FreeRTOS; is it because those fields get assigned by querying RTOS data structures at runtime?  (As opposed to being parameters provided in OpenOCD init scripts etc?

timsifive replied:
That's exactly why. Specifically, FreeRTOS_params is defined as const, and it didn't seem right to make it writable.
